// immutable-object-methods/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.chain = exports.map = exports.without = exports.set = exports.getIn = exports.setIn = exports.mergeDeep = exports.assign = undefined;

var _assign2 = require('./assign');

var _assign3 = _interopRequireDefault(_assign2);

var _mergeDeep2 = require('./merge-deep');

var _mergeDeep3 = _interopRequireDefault(_mergeDeep2);

var _setIn2 = require('./set-in');

var _setIn3 = _interopRequireDefault(_setIn2);

var _getIn2 = require('./get-in');

var _getIn3 = _interopRequireDefault(_getIn2);

var _set2 = require('./set');

var _set3 = _interopRequireDefault(_set2);

var _without2 = require('./without');

var _without3 = _interopRequireDefault(_without2);

var _map2 = require('./map');

var _map3 = _interopRequireDefault(_map2);

var _chain = require('./chain');

var _chain2 = _interopRequireDefault(_chain);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var assign = exports.assign = _assign3.default;
var mergeDeep = exports.mergeDeep = _mergeDeep3.default;
var setIn = exports.setIn = _setIn3.default;
var getIn = exports.getIn = _getIn3.default;
var set = exports.set = _set3.default;
var without = exports.without = _without3.default;
var map = exports.map = _map3.default;
var chain = exports.chain = (0, _chain2.default)({
  assign: assign, mergeDeep: mergeDeep, setIn: setIn, getIn: getIn, set: set, without: without, map: map
});

exports.default = {
  assign: assign,
  getIn: getIn,
  setIn: setIn,
  mergeDeep: mergeDeep,
  set: set,
  without: without,
  map: map,
  chain: chain
};
